dask/blockwise.py

```diff
@@ -181,7 +181,16 @@ def __init__(
             (name, tuple(ind) if ind is not None else ind) for name, ind in indices
         )
         self.numblocks = numblocks
-        self.concatenate = concatenate
+        # optimize_blockwise won't merge where `concatenate` doesn't match, so
+        # enforce a canonical value if there are no axes for reduction.
+        if all(
+            all(i in output_indices for i in ind)
+            for name, ind in indices
+            if ind is not None
+        ):
+            self.concatenate = None
```

Comment:
For my own understanding, we're deliberately ignoring the `concatenate` providing by the calling function here? That seems a bit strange, but I likely don't understand the context. Can you give an example where the calling function's concatenate should be ignored? Does it make sense to ask / require the caller to do this check?

Reply:
`concatenate` doesn't matter when there is no reduction happening, but some code passes `True` (e.g. `map_blocks`) and some doesn't (elementwise ops). I considered tweaking `map_blocks` to detect when there is no reduction (which I think is just when `drop_axes` is not used), but I don't know if there are other cases (and possibly user code) so this approach handles them all.

Possibly this could also be handled in `optimize_blockwise` (detect that mismatched `concatenate` doesn't matter because there is no reduction and allow fusion), but the interface to Blockwise doesn't make it particularly easy to determine.

Reply:
OK, thanks. I'll trust your instinct that this is easier to detect here than in the calling function!